Fetch (all) resources when building from source

MikeMcQuaid opened the pull request:
Also:
- when `brew test` or `brew postinstall` is run allow `Resource#stage`
  to fetch the resource.
- make `Formula#fetch` and `Resource#fetch` fetch external patches too.

Follow-up from #7549 and #7546.

- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Bo98 commented:
At some point, I think `prepare_patches` should be a part of the fetch logic given it does just that for patches.

MikeMcQuaid commented:
@Bo98 Not sure if you read the CI output or guessed correctly but: the lack of that was failing CI so I've added it in here and updated the description.

Bo98 commented:
No I didn't read the CI. I didn't actually expect your original commit to result in any behaviour change here relating to patches, so I'm surprised it did.

Bo98 commented on Library/Homebrew/resource.rb:
```diff
@@ -115,6 +115,8 @@ def files(*files)
   def fetch(verify_download_integrity: true)
     HOMEBREW_CACHE.mkpath
 
+    fetch_patches
+
     begin
       downloader.fetch
     rescue ErrorDuringExecution, CurlDownloadStrategyError => e
```
We need to make sure the formula installer never invokes `Formula.downloader.fetch` directly (I believe it does currently if I recall correctly) as doing that bypasses this override.

MikeMcQuaid replied:
Done!

Bo98 commented on Library/Homebrew/formula.rb:
```diff
@@ -1162,9 +1164,10 @@ def patch
   # @private
   def brew
     @prefix_returns_versioned_prefix = true
+    active_spec.fetch
```
So I can follow: in what scenarios is this needed? This can cause double fetching (albeit with caching in the second run).

MikeMcQuaid replied:
It's been made conditional. It's mainly needed for tests.

MikeMcQuaid commented:
@Bo98 CI is 🟢 now, let me know what you think.
